node/src/types/block/past_finality_signatures.rs

```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::PastFinalitySignatures;
+    use casper_types::{
+        bytesrepr::{FromBytes, ToBytes},
+        testing::TestRng,
+        PublicKey,
+    };
+    use std::collections::BTreeSet;
+
+    #[test]
+    fn empty_signatures() {
+        let rng = &mut TestRng::new();
+        let validators: Vec<_> = std::iter::repeat_with(|| PublicKey::random(rng))
+            .take(7)
+            .collect();
+        let original_signed = BTreeSet::new();
+
+        let past_finality_signatures =
+            PastFinalitySignatures::from_validator_set(&original_signed, validators.iter());
+
+        assert_eq!(past_finality_signatures.0, &[0]);
+
+        let signed = past_finality_signatures.into_validator_set(validators.into_iter());
+
+        assert_eq!(original_signed, signed);
+    }
+
+    #[test]
+    fn from_and_to_methods_match_in_a_simple_case() {
+        let rng = &mut TestRng::new();
+        let validators: Vec<_> = std::iter::repeat_with(|| PublicKey::random(rng))
+            .take(11)
+            .collect();
+        let signed = {
+            let mut signed = BTreeSet::new();
+            signed.insert(validators[2].clone());
+            signed.insert(validators[5].clone());
+            signed.insert(validators[6].clone());
+            signed.insert(validators[8].clone());
+            signed.insert(validators[10].clone());
+            signed
+        };
+
+        let past_finality_signatures =
+            PastFinalitySignatures::from_validator_set(&signed, validators.iter());
+
+        assert_eq!(past_finality_signatures.0, &[0b0010_0110, 0b1010_0000]);
+
+        let signed_ = past_finality_signatures.into_validator_set(validators.into_iter());
+
+        assert_eq!(signed, signed_);
+    }
+
+    #[test]
+    fn simple_serialization_roundtrip() {
+        let data = PastFinalitySignatures(vec![1, 2, 3, 4, 5]);
+
+        let serialized = data.to_bytes().unwrap();
+        assert_eq!(serialized.len(), data.0.len() + 4);
+        assert_eq!(data.serialized_length(), data.0.len() + 4);
+
+        let deserialized = PastFinalitySignatures::from_bytes(&serialized).unwrap().0;
+
+        assert_eq!(data, deserialized);
+    }
+
+    #[test]
+    fn serialization_roundtrip_of_empty_data() {
+        let data = PastFinalitySignatures::default();
+
+        let serialized = data.to_bytes().unwrap();
+        assert_eq!(serialized, &[0; 4]);
+        assert_eq!(data.serialized_length(), 4);
+
+        let deserialized = PastFinalitySignatures::from_bytes(&serialized).unwrap().0;
+
+        assert_eq!(data, deserialized);
+    }
+}
+
+impl crate::utils::specimen::LargestSpecimen for PastFinalitySignatures {
+    fn largest_specimen<E: crate::utils::specimen::SizeEstimator>(
+        estimator: &E,
+        _cache: &mut crate::utils::specimen::Cache,
+    ) -> Self {
+        PastFinalitySignatures(vec![u8::MAX; estimator.parameter("validator_count")])
```

Comment:
The length here should be validator count divided by 8 - we can fit 8 validators per byte in `PastFinalitySignatures`.

Reply:
Done